tools/slang-capability-generator/capability-generator-main.cpp

```diff
@@ -43,17 +50,48 @@ struct SerializedArrayView
     Index count;
 };
 
+struct CapabilitySharedContext
+{
+    CapabilityDef* ptrOfTarget = nullptr;
+    CapabilityDef* ptrOfStage = nullptr;
+};
+
+static void _removeFromOtherAtomsNotInThis(HashSet<const CapabilityDef*> thisSet, HashSet<const CapabilityDef*> otherSet, List<const CapabilityDef*> atomsToRemove)
+{
+    atomsToRemove.clear();
+    atomsToRemove.reserve(otherSet.getCount());
+    for (auto abstractAtom : otherSet)
+    {
+        if (thisSet.contains(abstractAtom))
+            continue;
+        atomsToRemove.add(abstractAtom);
+    }
+
+    for (auto atomToRemove : atomsToRemove)
+        otherSet.remove(atomToRemove);
+}
+
 struct CapabilityDef : public RefObject
 {
-    Index enumValue;
+public:
+
+    CapabilityDef(CapabilitySharedContext& context) : sharedContext(context)
+    {
+    }
+
     String name;
+    Index enumValue;
     CapabilityDisjunctionExpr expr;
     CapabilityFlavor flavor;
-    int rank;
+    int rank = 0; // optional, 0 is default
     List<List<CapabilityDef*>> canonicalRepresentation;
     SerializedArrayView serializedCanonicalRepresentation;
+    SourceLoc sourceLoc;
+    HashSet<const CapabilityDef*> abstractAtomsPresent;
 
-    CapabilityDef* getAbstractBase()
+    CapabilitySharedContext& sharedContext;
```

Comment:
`&` is not needed.
Just make a copy.

You seem to be abusing `&` too much.
Please use it more precisely.

When you make the variable `&`, it is basically a memory pointer that points to a variable outside of the class or function.
The life time of the variable can end while this class/function still uses it.
It is unnecessary to use `&` for this case and many other cases as a function parameter.

Reply:
`CapabilitySharedContext& sharedContext;` is a reference because imagine the case: `this->sharedContext.var = nullptr //invalidate pointer, not the data inside`

This case would not work if we copied the context, we need a pointer/reference to it (reference cannot be null and must be initialized, so ref is preferred)

Reply:
If there is a preferred method, I would be happy to change the implementation .